chain-signatures/node/src/protocol/message/filter.rs

```diff
@@ -0,0 +1,57 @@
+use std::num::NonZeroUsize;
+
+use tokio::sync::mpsc::{self, error::TryRecvError};
+
+use super::types::{MessageFilterId, Protocols};
+
+/// Maximum size for the filter of messages. This is roughly determined by the
+/// max number of protocols that can be within our system. It's not an upper
+/// bound but merely to serve as a good enough amount to maintain the IDs of
+/// protocols long enough on the case that they make it back into the system
+/// somehow after being erased.
+pub const MAX_FILTER_SIZE: NonZeroUsize = NonZeroUsize::new(64 * 1024).unwrap();
+
+#[derive(Debug)]
+pub(crate) struct MessageFilter {
+    filter_rx: mpsc::Receiver<(Protocols, u64)>,
+    filter: lru::LruCache<(Protocols, u64), ()>,
+}
+
+impl MessageFilter {
+    pub fn new(filter_rx: mpsc::Receiver<(Protocols, u64)>) -> Self {
+        Self {
+            filter_rx,
+            filter: lru::LruCache::new(MAX_FILTER_SIZE),
+        }
+    }
+
+    pub fn contains<M: MessageFilterId>(&mut self, msg: &M) -> bool {
+        // Check if the message is already in the filter. Doing `get` here will also
+        // update the LRU cache and promote the rank of this id to be most recent.
+        self.filter.get(&(M::PROTOCOL, msg.id())).is_some()
+    }
+
+    pub fn extend(&mut self) {
```

Comment:
nit: `extend` seems like an odd choice of a name. In Rust, I would personally only use it to do what [std::iter::Extend](https://doc.rust-lang.org/std/iter/trait.Extend.html#tymethod.extend) does.

I would maybe call it `update`, `refresh`, or if you like the extend word, `extend_from_receiver`. But I think using the word extend in such a way without a paramter is against common practice in Rust.

Obviously a matter of opinion, though, feel free to disagree and keep your naming :)